include/sparrow/typed_array.hpp

```diff
@@ -0,0 +1,358 @@
+         *
+         * @return An iterator to the end.
+         */
+        iterator end();
+        const_iterator cend() const;
+        ///@}
+
+        /*
+         * @brief Returns a range of the bitmap.
+         *
+         * @return A range of the bitmap.
+         */
+        const_bitmap_range bitmap() const;
+
+        /*
+         * @brief Returns a range of the values.
+         *
+         * @return A range of the values.
+         */
+        const_value_range values() const;
+
+        // Capacity
+
+        /*
+         * @brief Checks whether the container is empty.
+         *
+         * @description Returns whether the container is empty (i.e. whether its size is 0).
+         *
+         * @return true if the container is empty, false otherwise.
+         */
+        bool empty() const;
+
+        /*
+         * @brief Returns the number of elements.
+         *
+         * @description Returns the number of elements in the container, i.e. std::distance(begin(), end()).
+         *
+         * @return The number of elements in the container.
+         */
+        size_type size() const;
+
+        // TODO: Add reserve, capacity, shrink_to_fit
+
+        // Modifiers
+
+        // TODO: Implement insert, erase, push_back, pop_back, clear, resize, swap
+
+        std::partial_ordering operator<=>(const typed_array<T>& other) const;
+
+        bool operator==(const typed_array<T>& other) const;
+
+        bool operator!=(const typed_array<T>& other) const;
+
+        bool operator<(const typed_array<T>& other) const;
+
+        bool operator>(const typed_array<T>& other) const;
+
+        bool operator<=(const typed_array<T>& other) const;
+
+        bool operator>=(const typed_array<T>& other) const;
+
+    private:
+
+        array_data m_data;
+        layout_type m_layout;
+    };
+
+    // Constructors
+    template <class T, class L>
+    typed_array<T, L>::typed_array(array_data data)
+        : m_data(std::move(data))
+        , m_layout(m_data)
+    {
+    }
+
+    // Element access
+
+    template <class T, class L>
+    auto typed_array<T, L>::at(size_type i) -> reference
+    {
+        if (i >= size())
+        {
+            throw std::out_of_range(
+                "typed_array::at: index out of range for array of size " + std::to_string(size())
+                + " at index " + std::to_string(i)
+            );
+        }
+        return m_layout[i];
+    }
+
+    template <class T, class L>
+    auto typed_array<T, L>::at(size_type i) const -> const_reference
+    {
+        if (i >= size())
+        {
+            throw std::out_of_range(
+                "typed_array::at: index out of range for array of size " + std::to_string(size())
+                + " at index " + std::to_string(i)
+            );
+        }
+        return m_layout[i];
+    }
+
+    template <class T, class L>
+    auto typed_array<T, L>::operator[](size_type i) -> reference
+    {
+        return m_layout[i];
```

Comment:
Consider adding an assert here and in the following methods.

Reply:
Yes, set the asserts matching the preconditions and type-invariants 👍🏽 .